zyte_common_items/adapter.py

```diff
@@ -0,0 +1,92 @@
+
+    *value* is assumed not to be a mapping, which should be treated as a
+    non-empty value, but this function would treat as an empty value.
+    """
+    return value is None or (not value and not isinstance(value, (bytes, str)) and isinstance(value, Collection))
+
+
+class ZyteItemAdapter(AttrsAdapter):
+    """Adapter class for the itemadapter library that improves interaction with
+    :class:`zyte_common_items.Item` for itemadapter users like Scrapy.
+
+    Configure it as early as possible in your code::
+
+        from itemadapter import ItemAdapter
+        from zyte_common_items import ZyteItemAdapter
+
+        ItemAdapter.ADAPTER_CLASSES.appendleft(ZyteItemAdapter)
+
+    This adapter:
+
+    -   Allows interaction and serialization of fields not yet defined in the
+        corresponding zyte-common-items item but received as input from Zyte
+        Data API.
+
+    -   Removes keys with empty values from the output of
+        ``ItemAdapter.asdict()``, for a cleaner output.
+    """
+
+    @classmethod
+    def is_item(cls, item: Any) -> bool:
+        return isinstance(item, Item)
+
+    def get_field_meta(self, field_name: str) -> MappingProxyType:
+        if field_name in self._fields_dict:
+            return self._fields_dict[field_name].metadata  # type: ignore
+        elif field_name in self.item._unknown_fields_dict:
+            return MappingProxyType({})
+        raise KeyError(field_name)
+
+    def field_names(self) -> KeysView:
+        return KeysView({**self._fields_dict, **self.item._unknown_fields_dict})
+
+    def __getitem__(self, field_name: str) -> Any:
+        if field_name in self._fields_dict:
+            return getattr(self.item, field_name)
+        elif field_name in self.item._unknown_fields_dict:
+            return self.item._unknown_fields_dict[field_name]
+        raise KeyError(field_name)
+
+    def __setitem__(self, field_name: str, value: Any) -> None:
+        if field_name in self._fields_dict:
+            setattr(self.item, field_name, value)
+        else:
+            self.item._unknown_fields_dict[field_name] = value
+
+    def __delitem__(self, field_name: str) -> None:
+        if field_name in self._fields_dict:
+            del self._fields_dict[field_name]
+            delattr(self.item, field_name)
+        elif field_name in self.item._unknown_fields_dict:
+            del self.item._unknown_fields_dict[field_name]
+        else:
+            raise KeyError(
+                f"Object of type {self.item.__class__.__name__} does not contain a field with name {field_name}"
+            )
+
+    def __iter__(self) -> Iterator:
+        fields = [attr for attr in self._fields_dict if not _is_empty(getattr(self.item, attr))]
+        fields.extend(
+            attr
+            for attr in self.item._unknown_fields_dict
+            if (attr not in fields and not _is_empty(self.item._unknown_fields_dict[attr]))
```

Comment:
Why is `attr not in fields` check needed, is it an additional safety measure? I notice it because of a possible performance impact, as a list lookup.

It's hard to imagine a scenario where `fields` list is large, so the overhead is small, but it still may worth it to make it a set lookup. It's always a worst-case lookup here, the whole fields need to be iterated.

Actually, I'm also not sure how fields.extend handles the extension - if `fields` is changed on each iteration, the `not in fields` lookup makes the loop quadratic.

Some numbers, to demonstrate that my comment is not very important :) 
```py
In [1]: lst = list("abcdefsldjfgh")

In [2]: lst
Out[2]: ['a', 'b', 'c', 'd', 'e', 'f', 's', 'l', 'd', 'j', 'f', 'g', 'h']

In [3]: lst_set = set(lst)

In [4]: %timeit 'A' in lst
238 ns ± 0.708 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)

In [5]: %timeit 'A' in lst_set
30.7 ns ± 0.0337 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)

In [6]: %timeit set(lst)
265 ns ± 0.121 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
```